craft_application/models/project.py

```diff
@@ -93,14 +96,122 @@ class BuildPlannerConfig(CraftBaseConfig):
     """The BuildPlanner model uses attributes from the project yaml."""
 
 
+class Platform(CraftBaseModel):
+    """Project platform definition."""
+
+    build_on: list[str] | None = pydantic.Field(min_items=1, unique_items=True)
+    build_for: list[str] | None = pydantic.Field(
+        min_items=1, max_items=1, unique_items=True
+    )
+
+    @pydantic.validator(  # pyright: ignore[reportUnknownMemberType,reportUntypedFunctionDecorator]
+        "build_on", "build_for"
+    )
+    def _validate_architectures(cls, values: list[str]) -> list[str]:
+        """Validate the architecture entries."""
+        for architecture in values:
+            if not is_valid_architecture(architecture):
+                raise errors.CraftValidationError(
+                    f"Invalid architecture: {architecture!r} "
+                    "must be a valid debian architecture."
+                )
+
+        return values
+
+    @pydantic.root_validator(  # pyright: ignore[reportUntypedFunctionDecorator,reportUnknownMemberType]
+        skip_on_failure=True
+    )
+    @classmethod
+    def _validate_platform_set(cls, values: Mapping[str, Any]) -> Mapping[str, Any]:
+        """If build_for is provided, then build_on must also be."""
+        if not values.get("build_on") and values.get("build_for"):
+            raise errors.CraftValidationError(
+                "'build_for' expects 'build_on' to also be provided."
+            )
+
+        return values
+
+
+def _get_default_platform() -> dict[str, Platform]:
+    return {
+        get_host_architecture(): Platform(
+            build_on=[get_host_architecture()],
+            build_for=[get_host_architecture()],
+        )
+    }
+
+
+def _validate_platforms(platforms: dict[str, Platform]) -> dict[str, Platform]:
+    """Validate that platforms are valid and populate platform entries.
+
+    :param platforms: The platforms to validate.
+
+    :raises CraftValidationError: If a platform does not have a `build-on` and
+    `build-for` and the platform entry's name is not a valid debian architecture.
+
+    :returns: The dict of platforms with populated entries.
+    """
+    for platform_entry, platform in platforms.items():
+        if not platform:
+            if not is_valid_architecture(platform_entry):
+                raise errors.CraftValidationError(
+                    f"Invalid platform: {platform_entry!r} must either be a "
+                    "valid debian architecture or define architectures with "
+                    "'build-on' and 'build-for'."
+                )
+            # populate "empty" platforms entries
+            platforms[platform_entry] = Platform(
```

Comment:
I might try making this a dictionary instead, this is creating a weird mix of Platforms and Dictionary objects when loading yaml.

And if I do that, then maybe I will drop the validation of architectures as well.